watchtower/wtclient/queue_test.go

```diff
@@ -0,0 +1,305 @@
+package wtclient
+
+import (
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btclog"
+	"github.com/lightningnetwork/lnd/lntest/wait"
+	"github.com/lightningnetwork/lnd/watchtower/wtdb"
+	"github.com/lightningnetwork/lnd/watchtower/wtmock"
+	"github.com/stretchr/testify/require"
+)
+
+// TestDiskOverflowQueue tests that the DiskOverflowQueue behaves as expected.
+func TestDiskOverflowQueue(t *testing.T) {
+	t.Parallel()
+
+	const (
+		maxInMemItems = 5
+		waitTime      = time.Second * 2
+	)
+
+	getNext := func(q *DiskOverflowQueue[*wtdb.BackupID],
+		i int) *wtdb.BackupID {
+
+		var item *wtdb.BackupID
+		select {
+		case item = <-q.NextBackupID():
+		case <-time.After(waitTime):
+			t.Fatalf("task %d not received in time", i)
+		}
+
+		return item
+	}
+
+	enqueue := func(q *DiskOverflowQueue[*wtdb.BackupID],
+		task *wtdb.BackupID) {
+
+		err := q.QueueBackupID(task)
+		require.NoError(t, err)
+	}
+
+	waitForNumDisk := func(db wtdb.Queue[*wtdb.BackupID], num int) {
+		err := wait.Predicate(func() bool {
+			n, err := db.Len()
+			require.NoError(t, err)
+
+			return n == uint64(num)
+		}, waitTime)
+		require.NoError(t, err)
+	}
+
+	assertNumDisk := func(db wtdb.Queue[*wtdb.BackupID], num int) {
+		n, err := db.Len()
+		require.NoError(t, err)
+		require.EqualValues(t, num, n)
+	}
+
+	t.Run("overflow to disk", func(t *testing.T) {
+		t.Parallel()
+
+		// Generate some backup IDs that we want to add to the queue.
+		tasks := genBackupIDs(10)
+
+		// Create a mock db.
+		db := wtmock.NewQueueDB[*wtdb.BackupID]()
+
+		// New mock logger.
+		log := newMockLogger(t.Logf)
+
+		// Init the queue with the mock DB.
+		q, err := NewDiskOverflowQueue[*wtdb.BackupID](
+			db, maxInMemItems, log,
+		)
+		require.NoError(t, err)
+
+		// Start the queue.
+		require.NoError(t, q.Start())
+
+		// Initially there should be no items on disk.
+		assertNumDisk(db, 0)
+
+		// Start filling up the queue.
+		enqueue(q, tasks[0])
+		enqueue(q, tasks[1])
+		enqueue(q, tasks[2])
+		enqueue(q, tasks[3])
+		enqueue(q, tasks[4])
+
+		// The queue should now be full, so any new items should be
+		// persisted to disk.
+		enqueue(q, tasks[5])
+		waitForNumDisk(db, 1)
+
+		// Now pop all items from the queue to ensure that the item
+		// from disk is loaded in properly once there is space.
+		require.Equal(t, tasks[0], getNext(q, 0))
+		require.Equal(t, tasks[1], getNext(q, 1))
```

Comment:
Enqueuing a 7th item here when in `diskMode` could check that there's no out of order processing